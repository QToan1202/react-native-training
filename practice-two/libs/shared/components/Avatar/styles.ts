import { Image } from 'react-native'
import { GetProps, Stack, styled } from 'tamagui'

const StyledStack = styled(Stack, {
  space: 10,
  alignItems: 'center',
  columnGap: 4,

  variants: {
    inline: {
      true: {
        flexDirection: 'row',
      },
    },

    block: {
      true: {
        flexDirection: 'column',
      },
    },
  } as const,

  defaultVariants: {
    inline: true,
  },
})

export type StyledStackProps = GetProps<typeof StyledStack>

export const StyledImage = styled(Image, {
  name: 'Image',
  source: { uri: '' },
})

export type StyledImageProps = GetProps<typeof StyledImage>

export default StyledStack
