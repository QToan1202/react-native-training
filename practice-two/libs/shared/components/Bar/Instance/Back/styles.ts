import { StyleSheet } from 'react-native'

const styles = StyleSheet.create({
  container: {
    paddingBottom: 10,
  },
})

export default styles
