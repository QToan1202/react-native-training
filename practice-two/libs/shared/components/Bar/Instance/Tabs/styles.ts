import { StyleSheet } from 'react-native'

import { COLORS } from '@constants'

const styles = StyleSheet.create({
  shadow: {
    shadowColor: COLORS.BLACK,
    shadowOffset: {
      width: 0,
      height: 12,
    },
    shadowOpacity: 0.6,
    shadowRadius: 16,
  },
})

export default styles
