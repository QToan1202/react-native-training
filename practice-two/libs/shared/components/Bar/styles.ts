import { StyleSheet } from 'react-native'

import { COLORS } from '../../constants'

const styles = StyleSheet.create({
  bar: {
    backgroundColor: COLORS.PRIMARY,
    rowGap: 24,
  },
  title: {
    textTransform: 'capitalize',
  },
  backBtn: {
    position: 'absolute',
    left: 0,
    top: 0,
    zIndex: 1,
  },
  itemsContainer: {
    justifyContent: 'space-between',
  },
})

export default styles
