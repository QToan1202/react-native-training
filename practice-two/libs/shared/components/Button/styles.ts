import {
  ButtonIcon,
  GetProps,
  SizableText,
  Stack,
  createStyledContext,
  styled,
  withStaticProperties,
} from 'tamagui'

const ButtonContext = createStyledContext({
  color: undefined,
  ellipse: undefined,
  fontFamily: undefined,
  fontSize: undefined,
  fontStyle: undefined,
  fontWeight: undefined,
  letterSpacing: undefined,
  maxFontSizeMultiplier: undefined,
  size: undefined,
  textAlign: undefined,
  lineHeight: undefined,
})

const CustomButtonFrame = styled(Stack, {
  name: 'Button',
  context: ButtonContext,
  paddingVertical: '$space.3',
  borderRadius: '$radius.12',
  borderWidth: '$space.1',
  borderColor: '$color.transparent',
  display: 'flex',
  flexDirection: 'row',
  justifyContent: 'center',
  alignItems: 'center',
  columnGap: 6, // getTokens().space[2].val - 1
  pressStyle: {
    opacity: 0.5,
  },

  variants: {
    variant: {
      primary: {
        borderColor: '$color.primary',
        backgroundColor: '$color.primary',
        color: '$color.white',
      },

      secondary: {
        borderColor: '$color.transparent',
        backgroundColor: '$color.secondary',
        color: '$color.primary',
      },

      tertiary: {
        borderColor: '$color.white',
        backgroundColor: '$color.transparent',
        color: '$color.white',
      },

      quaternary: {
        borderRadius: '$radius.0',
        borderColor: '$color.white',
        backgroundColor: '$color.white',
        color: '$color.black',
      },
    },

    shrink: {
      true: {
        alignSelf: 'flex-start',
        paddingVertical: '$space.2',
        paddingHorizontal: 6,
      },
    },

    isDisable: {
      true: {
        opacity: 0.5,
      },
    },
  } as const,

  defaultVariants: {
    variant: 'primary',
  },
})

const ButtonText = styled(SizableText, {
  name: 'Button',
  context: ButtonContext,
  userSelect: 'none',

  variants: {
    default: {
      true: {
        fontWeight: '$2',
        textTransform: 'capitalize',
        textAlign: 'center',
        lineHeight: '$2',
      },
    },
    size: {
      '...fontSize': (name, { font }) => ({
        fontSize: font?.size[name],
      }),
    },
  } as const,

  defaultVariants: {
    default: true,
  },
})

const ButtonStyled = withStaticProperties(CustomButtonFrame, {
  Text: ButtonText,
  Icon: ButtonIcon,
  Props: ButtonContext.Provider,
})

export type ButtonProps = GetProps<typeof ButtonText> & GetProps<typeof CustomButtonFrame>

export default ButtonStyled
