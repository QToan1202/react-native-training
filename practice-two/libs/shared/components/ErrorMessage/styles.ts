import { StyleSheet } from 'react-native'

import { COLORS } from '../../constants'

const styles = StyleSheet.create({
  errorWrapper: {
    height: 20,
  },
  error: {
    paddingLeft: 20,
    color: COLORS.RED,
    fontSize: 13,
    lineHeight: 20,
  },
})

export default styles
