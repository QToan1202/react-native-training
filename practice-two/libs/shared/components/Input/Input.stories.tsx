import { Alert, StyleSheet, View } from 'react-native'
import { SubmitHandler, useForm } from 'react-hook-form'
import { ComponentMeta, ComponentStory } from '@storybook/react-native'
import { getTokens } from 'tamagui'

import { IForm } from '@types'

import { Button, Input } from '../index'
import { InputProps } from './index'

export default {
  title: 'components/Input',
  component: Input,
} as ComponentMeta<typeof Input>

const styles = StyleSheet.create({
  container: {
    padding: 8,
    display: 'flex',
    rowGap: 7,
  },
})

const Template: ComponentStory<typeof Input> = (arg: Omit<InputProps, 'control'>) => {
  const { control, handleSubmit } = useForm<IForm>()
  const onSubmit: SubmitHandler<IForm> = (data) => {
    Alert.alert(JSON.stringify(data))
  }

  return (
    <View style={styles.container}>
      <Input control={control} {...arg} />
      <Button title="Submit" onPress={handleSubmit(onSubmit)} />
    </View>
  )
}

export const Default = Template.bind({})
Default.args = {
  name: 'email',
  placeholder: 'Enter your email',
  placeholderTextColor: getTokens().color.white.val,
}

export const InputWithLabel = Template.bind({})
InputWithLabel.args = {
  label: 'First name',
  name: 'firstName',
  placeholder: 'Enter your name',
  placeholderTextColor: getTokens().color.black.val,
}
