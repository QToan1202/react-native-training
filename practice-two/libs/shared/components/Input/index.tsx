import { forwardRef, memo, useCallback, useState } from 'react'
import isEqual from 'react-fast-compare'
import { Control, UseControllerProps, useController } from 'react-hook-form'
import { Stack, StackProps } from 'tamagui'
import { TextInput } from 'react-native'

import { ErrorMessage, IForm } from '@practice-two/shared'

import { Label, StyledInput, StyledInputProps } from './styles'

export type InputProps = StyledInputProps & {
  name: keyof IForm
  control: Control<IForm>
  label?: string
  rules?: UseControllerProps['rules']
  isShowError?: boolean
  containerStyle?: StackProps
}

const Input = forwardRef<TextInput, InputProps>(
  ({ label, name, control, rules, isShowError = false, containerStyle, ...rest }, ref) => {
    const {
      field,
      formState: { errors },
    } = useController<IForm>({
      control,
      defaultValue: '',
      name,
      rules,
    })
    const [isInputFocus, setIsInputFocus] = useState<boolean>(false)
    const handleFocusInput = useCallback(() => setIsInputFocus(true), [])
    const handleBlurInput = useCallback(() => setIsInputFocus(false), [])

    return (
      <Stack w="100%" pos="relative" {...containerStyle}>
        {label && <Label forceStyle={isInputFocus ? 'focus' : 'press'}>{label}</Label>}
        <StyledInput
          ref={ref}
          hasLabel={!!label}
          value={String(field.value)}
          onChangeText={field.onChange}
          onFocus={handleFocusInput}
          onBlur={handleBlurInput}
          {...rest}
        />
        {isShowError && <ErrorMessage pos="absolute" bottom={-20} left={10} error={errors[name]} />}
      </Stack>
    )
  }
)

export default memo(Input, isEqual)
