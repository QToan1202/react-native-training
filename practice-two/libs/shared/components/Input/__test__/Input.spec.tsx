import { useForm } from 'react-hook-form'
import { render, fireEvent, renderHook } from '@testing-library/react-native'

import { IForm } from '@types'

import Input from '../index'
import styles from '../styles'

describe('Testing Input component', () => {
  it('Render Input component success', () => {
    const { result } = renderHook(() => useForm<IForm>())
    const { control } = result.current
    const component = render(
      <Input control={control} name="firstName" label="first name" />
    ).toJSON()

    expect(component).toMatchSnapshot()
  })

  it('Input can change text', () => {
    const { result } = renderHook(() => useForm<IForm>())
    const { control } = result.current
    const { getByPlaceholderText } = render(
      <Input control={control} name="firstName" placeholder="first name" />
    )
    const element = getByPlaceholderText('first name')

    fireEvent.changeText(element, 'new values')
    expect(element.props.value).toBe('new values')
  })

  it('Input have the right styles when focus or blur', () => {
    const { result } = renderHook(() => useForm<IForm>())
    const { control } = result.current
    const { getByPlaceholderText } = render(
      <Input control={control} label="first name" name="firstName" placeholder="first name" />
    )
    const element = getByPlaceholderText('first name')

    fireEvent(element, 'focus')
    expect(element).toHaveStyle(styles.inputFocus)

    fireEvent(element, 'blur')
    expect(element).toHaveStyle(styles.inputWithLabel)
  })
})
