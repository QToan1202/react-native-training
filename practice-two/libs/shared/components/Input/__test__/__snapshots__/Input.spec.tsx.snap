// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Input component Render Input component success 1`] = `
<View>
  <Text
    style={
      [
        {
          "color": "#fff",
          "fontFamily": "Montserrat_400Regular",
        },
        {
          "fontSize": 14,
        },
        {
          "color": "#3a2c3c",
          "fontFamily": "Montserrat_500Medium",
        },
      ]
    }
  >
    first name
  </Text>
  <TextInput
    onBlur={[Function]}
    onChangeText={[Function]}
    onFocus={[Function]}
    style={
      [
        {
          "borderBottomColor": "#dbdbde",
          "borderBottomWidth": 0.5,
          "color": "#000",
          "fontFamily": "Montserrat_500Medium",
          "fontSize": 16,
          "paddingVertical": 5,
        },
        {},
        undefined,
      ]
    }
    value=""
  />
</View>
`;
